File: structural_design_patterns/src/main/java/org/example/structural/dto/BookDto.java
package org.example.structural.dto;

import lombok.Data;

@Data
public class BookDto {
    private Long id;
    private String title;
    private String author;
    private double price;
    private String theme;
    private Long ownerId;
    private boolean forSale;

    // Explicit getters and setters to avoid relying on Lombok at build time
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public String getTheme() {
        return theme;
    }

    public void setTheme(String theme) {
        this.theme = theme;
    }

    public Long getOwnerId() {
        return ownerId;
    }

    public void setOwnerId(Long ownerId) {
        this.ownerId = ownerId;
    }

    public boolean isForSale() {
        return forSale;
    }

    public void setForSale(boolean forSale) {
        this.forSale = forSale;
    }
}

File: flutter/flutter_application_1/lib/features/books/data/models/book_dto.dart
// lib/features/books/data/models/book_dto.dart

class BookDto {
  final int? id;
  final String title;
  final String author;
  final String theme;
  final double price;
  final int? ownerId;
  final bool forSale;

  BookDto({
    this.id,
    required this.title,
    required this.author,
    required this.theme,
    required this.price,
    this.ownerId,
    this.forSale = false,
  });

  factory BookDto.fromJson(Map<String, dynamic> json) {
    return BookDto(
      id: json['id'],
      title: json['title'],
      author: json['author'],
      theme: json['theme'],
      price: json['price'].toDouble(),
      ownerId: json['ownerId'],
      forSale: json['forSale'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'author': author,
      'theme': theme,
      'price': price,
      'ownerId': ownerId,
      'forSale': forSale,
    };
  }
}
